File: js/nqueen.js
#!/usr/bin/env k8

function nq_solve(n) {
	let a = [], l = [], c = [], r = [], m = 0;
	const y0 = (1<<n) - 1;
	for (let k = 0; k < n; ++k)
		a[k] = -1, l[k] = c[k] = r[k] = 0;
	for (let k = 0; k >= 0;) {
		const y = (l[k] | c[k] | r[k]) & y0;
		if ((y ^ y0) >> (a[k] + 1)) {
			let i;
			for (i = a[k] + 1; i < n; ++i)
				if ((y & 1<<i) == 0) break;
			if (k < n - 1) {
				const z = 1<<i;
				a[k++] = i;
				l[k] = (l[k-1]|z)<<1, c[k] = c[k-1]|z, r[k] = (r[k-1]|z)>>1;
			} else ++m, --k;
		} else a[k--] = -1;
	}
	return m;
}

var ccc = {
	print: typeof print == "function"? print : console.log,
	argv: typeof k8_version == "function"? arguments.slice(0) : typeof Deno == "object"? Deno.args.slice(0) : typeof Bun == "function"? Bun.argv.slice(2) : process.argv.splice(2)
};

function main(args) {
	let n = 15;
	if (args.length > 0) n = parseInt(args[0]);
	print(nq_solve(n));
}

main(ccc.argv);

File: src/ruby/nqueen.rb
# frozen_string_literal: true

def nq_solve(n)
  a = Array.new(n, -1)
  l = Array.new(n, 0)
  c = Array.new(n, 0)
  r = Array.new(n, 0)
  y0 = (1 << n) - 1
  m = 0
  k = 0

  while k >= 0
    y = (l[k] | c[k] | r[k]) & y0
    if (y ^ y0) >> (a[k] + 1) != 0
      i = a[k] + 1
      while i < n && (y & 1 << i) != 0
        i += 1
      end
      if k < n - 1
        z = 1 << i
        a[k] = i
        k += 1
        l[k] = (l[k-1] | z) << 1
        c[k] = c[k-1] | z
        r[k] = (r[k-1] | z) >> 1
      else
        m += 1
        k -= 1
      end
    else
      a[k] = -1
      k -= 1
    end
  end

  m
end

n = 15
if ARGV.length > 0
  n = ARGV[0].to_i
end

puts nq_solve(n)
